### PR TITLE
MM-55060 - Calls: Add PushTypeCalls for calls-specific push notifications

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- This is to support calls-specific push notifications, see also:
  - https://github.com/mattermost/mattermost-plugin-calls/pull/561
  - https://github.com/mattermost/mattermost/pull/25405
  - https://github.com/mattermost/mattermost-mobile/pull/7623
- This exposes a `/version` route right beside the existing `/root` route. It's not under the api because it's not part of the api (imo). I renamed the `git_hash` to `build_hash` which is more accurate (esp considering we're also showing the `git_version`, which is the version of _git_, not of the proxy, so pretty confusing). I also shortened it to the short hash because there's no real reason to expose the full hash. 
- The `/version` route will be requested once per calls plugin reload (in the `onActivate`).
-  This lets Calls be backward compatible with older push proxies and use correct typing in the future for calls push notification messages, which allows us to internationalize the call started push messages in calls v1.0.
- I tagged Agniva on these because he was doing the push-proxy recently, but please review these too @enahum if you wish.

- NOTE: will need to bump to 5.27.0 after this is merged for the Calls code to work.

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-55060

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -23,7 +25,7 @@ ifeq ($(DIFF), 1)
 endif
 
 PP_PKG=github.com/mattermost/mattermost-push-proxy/internal/version
-LDFLAGS="-X $(PP_PKG).gitVersion=$(GIT_VERSION) -X $(PP_PKG).gitCommit=$(GIT_HASH) -X $(PP_PKG).gitTreeState=$(GIT_TREESTATE) -X $(PP_PKG).buildDate=$(BUILD_DATE)"
+LDFLAGS="-X $(PP_PKG).gitVersion=$(GIT_VERSION) -X $(PP_PKG).buildHash=$(BUILD_HASH) -X $(PP_PKG).buildTagLatest=$(BUILD_TAG_LATEST) -X $(PP_PKG).buildTagCurrent=$(BUILD_TAG_CURRENT) -X $(PP_PKG).gitTreeState=$(GIT_TREESTATE) -X $(PP_PKG).buildDate=$(BUILD_DATE)"
```
Comment (agnivade) [inline Makefile:28]: Heads up that there are some changes to the Makefile here: https://github.com/mattermost/mattermost-push-proxy/pull/109.

File: server/server.go [hunk 2]
```diff
@@ -150,6 +151,22 @@ func root(w http.ResponseWriter, r *http.Request) {
 	_, _ = w.Write([]byte("<html><body>Mattermost Push Proxy</body></html>"))
 }
 
+func (s *Server) version(w http.ResponseWriter, _ *http.Request) {
```
Comment (agnivade) [inline server/server.go:154]: Can we have a unit test for this?

File: Makefile (hunk 1 above)
Comment (cpoile) [reply Makefile]: Thanks, merged and updated the Makefile.

File: server/server.go (hunk 2 above)
Comment (cpoile) [reply server/server.go:154]: Yep, done.

File: internal/version/version.go [hunk 3]
```diff
@@ -33,7 +37,8 @@ func GetVersion() error {
 
 type Info struct {
 	GitVersion   string
-	GitCommit    string
```
Comment (streamer45) [inline internal/version/version.go:36]: This is technically a breaking change. I assume we are not relying on it somewhere else?

File: server/server.go [hunk 4]
```diff
@@ -150,6 +151,22 @@ func root(w http.ResponseWriter, r *http.Request) {
...
+	info := version.VersionInfo()
+	ret := map[string]interface{}{
+		"version": info.BuildVersion,
+		"hash":    info.BuildHash,
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(ret); err != nil {
```
Comment (streamer45) [inline server/server.go:162]: Any reason for not marshaling the info struct? We could always omit fields through struct tags if needed but at least we'd avoid the map altogether.

File: internal/version/version.go (hunk 3 above)
Comment (cpoile) [reply internal/version/version.go:36]: It's only used to print details when starting up (and the version endpoint, of course)

File: server/server.go (hunk 4 above)
Comment (cpoile) [reply server/server.go:162]: I didn't want to publicly expose what would be considered internal details, and this is all we needed. Happy to marshal it all, but maybe I'll ask sec first?

Comment (streamer45) [reply server/server.go:162]: Just expose the ones we need for now and `json:"-"` the rest?

Comment (cpoile) [reply server/server.go:162]: Ah, right, missed that part of your response. :)

**Review by streamer45 — APPROVED** [review]
Nicely done!

**Comment by enahum** [comment]
This approach solves the new plugin + new proxy + new app, but what about the other cases?

For example what if the app is not up to date?

Perhaps we should have a quick convo about this if you want me to clarify.

In my mind, regardless of the combination of elements: server, proxy, plugin, app. The notification should work as was already introduced. Backwards compatibility needs to be kept to avoid bug reports

**Review by agnivade — APPROVED** [review]

**Comment by cpoile** [comment]
@enahum Excellent point, overlooked that. Fixed (see: https://github.com/mattermost/mattermost/pull/25405#issuecomment-1810989581 ) for an explanation. Thank you!

**Comment by enahum** [comment]
@cpoile thanks for the change.. now it seems to take care of the backwards compatibility

**Review by agnivade — APPROVED** [review]

**Comment by cpoile** [comment]
@streamer45 Nope, but I did want to keep it so that it's there in previous versions if we ever do need it in the future.

**Review by streamer45 — APPROVED** [review]

**Pull request merged**